compile-time-programming/requires-expressions.md

```diff
@@ -0,0 +1,117 @@
+# Module name: Requires Expressions
+
+## Overview
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational</td>
+    <td>Define and use requires-expressions to check satisfaction of expressions by given parameters</td>
+  </tr>
+  <tr>
+    <td>Main</td>
+    <td>Define and use requires-expressions to check properties of expressions</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td></td>
+  </tr>
+</table>
+
+## Motivation
+
+Requires-expressions allow a developer to perform compile-time evaluation 
+on the validity of other expressions. These are fundamental to the ability 
+to write concepts. [[Compile-time programming: concepts]][1]
+
+## Topic introduction
+
+Requires-expressions are compile-time predicates which evaluate to true 
+when their specified set of expressions are all valid for a given set of 
+inputs.
+
+## Foundational: Writing requires-expressions
+
+### Background/Required Knowledge
+
+A student is able to:
+
+* Write and use a function template [[Compile-time programming: function templates]][2]
+* Differentiate between expressions and statements
+
+It is helpful if:
+
+* A student is aware of the constraints imposed by their template definitions
+
+### Student outcomes
+
+A student should be able to:
+
+1. Write a simple-requirement to assert the validity of an expression
+2. Write a type-requirement to check the existence of a type by its identifier
+3. Write a compound-requirement to test the resulting type of an expression
+4. Write a nested-requirement to test the constexpr value of an operation, as opposed to just the syntactic validity
+5. Use a requires-expression within a concept, requires-clause, or `if constexpr` condition
+
+### Caveats
+
+To require expressions which evaluate to a boolean value, such as 
+`sizeof(t) == 4`, to only be satisfied when the result of the expression is 
+`true`, you must use a nested-requirement, such as 
+`requires sizeof(t) == 4;`. Omitting the `requires` results in a 
```

Comment:
```suggestion
To require that one or more expressions, which evaluate to a boolean value 
like `sizeof(t) == 4`, evaluate to `true` a nested-requirement is needed 
(e.g., `requires sizeof(t) == 4;`). Omitting the `requires` results in a 
```

I find the first sentence a bit hard to read/understand. This would be my attempt to reword it.

Reply:
I would lose the "one or more":
> To require that a particular boolean expression evaluate to `true`, a nested-requirement is needed (e.g., `requires sizeof(t)==4;`). Omitting the `requires` keyword here results in a...

I just filed https://bugs.llvm.org/show_bug.cgi?id=47941 about Clang's lack of diagnostic here, btw.

Reply:
:+1: I adapted the wording.